### ISSUE TITLE
Integrating Autofac

### ISSUE THREAD

**Issue by justsayno** [sstorie/experiments#4]
Hi there,

Don't know if you have used Autofac (or use some other IoC container) but I am having trouble using your example and getting DI working.

I have created a repo and a stack overflow question but I thought I'd come to you as your tutorial was great and its all working except for this last peice.

SO:

http://stackoverflow.com/questions/37912330/autofac-injection-of-ihubcontext-signalr-mvc

Repo:

https://github.com/justsayno/signalr-autofac

**Comment by sstorie** [comment]
Yes I have and I think I can provide you some code showing one way to do
this. I'm on my phone at the moment but when I get back to my laptop I'll
try write a quick example.

**Comment by sstorie** [comment]
*Ok, I provided an answer to your SO question too, but for completeness I'll post the same thing here too*

I've been able to do this, but it's not as clean as I would normally like. The primary issue here is that IHubContext does not reflect what specific *type* of hub it is...it's just a generic handle. So what I've done is to create a named registration within Autofac to register IHubContext using a specific SignalR hub:

    builder.Register(ctx => 
        ctx.Resolve<IDependencyResolver>()
           .Resolve<IConnectionManager>()
           .GetHubContext<EventHub>())
           .Named<IHubContext>("EventHub");

Then I set up specific registrations for the objects I'll be injecting this hub into. This could be either an ApiController, or perhaps some other service that is then injected into controllers using the standard Autofac/WebApi integration. This "specific" registration is the piece I don't like, but I don't know of a cleaner way around it. Here's what that would look like:

    builder.RegisterType<TaskController>()
           .WithParameter(
               new ResolvedParameter(
                   (pi, ctx) => pi.ParameterType == typeof(IHubContext),
                   (pi, ctx) => ctx.ResolveNamed<IHubContext>("EventHub")
           )
     );

Now Autofac should recognize that you want to inject IHubContext into the `TaskController` and provide the specific registration named *EventHub* when injecting it.

http://stackoverflow.com/a/37913821/571237

**Comment by justsayno** [comment]
I have accepted your answer because I think you are right. But I have another issue entirely I think because I tried just reverting back to using the GlobalHost method in your example and It doesn't work.

Just because you might be able to help and have been really helpful so far!

The issue I am having is that those events in the subscribe

```
public async Task Subscribe(string channel)
        {
            await Groups.Add(Context.ConnectionId, channel);

            var ev = new ChannelEvent
            {
                ChannelName = "admin",
                Name = "user.subscribed",
                Data = new
                {
                    Context.ConnectionId,
                    ChannelName = channel
                }
            };

            await Publish(ev);
        }
```

Get received just fine by my front end. But if I send the very same event with practically the same code from an API controller it doesn't get received.. I am at a loss as to what I'm doing different to your example.

My front-end is slightly more complicated than yours (using Angular2 + redux) but I successfully get the connection and CAN receive messages just not send from anywhere but the Hub subscribe method itself. So I feel that eliminates the chance it is a front end issue...

I'd be happy to put my code into Bit Bucket or something and share it with you if you would be awesome enough to have a look? Can't share it on here unfortunately.
